addon_imps/storage/gitlab.py

```diff
@@ -0,0 +1,182 @@
+from addon_toolkit.interfaces.storage import (
+    ItemResult,
+    ItemSampleResult,
+    ItemType,
+)
+
+
+FOLDER_ITEM_TYPES = frozenset(["subfolder", "tree", "folder"])
+
+
+class GitlabStorageImp(storage.StorageAddonHttpRequestorImp):
+    """storage on gitlab
+
+    see https://developers.google.com/drive/api/reference/rest/v3/
+    """
+
+    async def get_external_account_id(self, _: dict[str, str]) -> str:
+        async with self.network.GET("user/preferences") as response:
+            resp_json = await response.json_content()
+            return resp_json.get("user_id", "")
+
+    async def list_root_items(self, page_cursor: str = "") -> storage.ItemSampleResult:
+        query_params = self._page_cursor_or_query(
+            page_cursor,
+            {
+                "membership": "true",
+                "simple": "true",
+                "pagination": "true",
+                "sort": "asc",
+            },
+        )
+        async with self.network.GET("projects", query=query_params) as response:
+            resp = await response.json_content()
+            return ItemSampleResult(
+                items=[Repository.from_json(item).item_result for item in resp],
+                next_sample_cursor=self._get_next_cursor(response.headers),
+            )
+
+    def _get_next_cursor(self, headers):
+        next_link_candidates = [
+            item for item in headers["Link"].split(",") if 'rel="next"' in item
+        ]
+        if not next_link_candidates:
+            return None
+        next_link_candidate: str = next_link_candidates[0]
+
+        return (
+            next_link_candidate.strip()
+            .removeprefix("<")
+            .removesuffix('>; rel="next"')
+            .split("?", maxsplit=1)[1]
+        )
+
+    def _page_cursor_or_query(self, page_cursor: str, query: dict):
+        if page_cursor:
+            return dict(urllib.parse.parse_qsl(page_cursor))
+        else:
+            return query
+
+    async def get_item_info(self, item_id: str) -> storage.ItemResult:
+        parsed_id = ItemId.parse(item_id)
+        if parsed_id.file_path:
+            return await self.get_file_or_folder(parsed_id)
+
+        return await self._get_repository(parsed_id.repo_id)
+
+    async def _get_repository(self, repo_id):
+        async with self.network.GET(f"projects/{repo_id}") as response:
+            content = await response.json_content()
+            return Repository.from_json(content).item_result
+
+    async def get_file_or_folder(self, parsed_id: ItemId):
+        async with self.network.GET(f"projects/{parsed_id.repo_id}") as response:
+            content = await response.json_content()
+            ref = content.get("default_branch")
+        if file_item := await self._get_file(parsed_id, ref):
+            return file_item
+        # try to list files under folder, if it succeeds, proceed to return folder, else propagate the error
+        await self.list_child_items(parsed_id.raw_id)
```

Comment:
I think you can remove this comment

Reply:
I'd like to leave it to clarify why this call is being made